src/v/storage/disk_log_impl.cc

```diff
@@ -1731,6 +1734,474 @@ disk_log_impl::make_cached_reader(log_reader_config config) {
+      last,
+      model::boundary_type::inclusive,
+      io_priority);
+
+    // compute size
+    size_t total_size = 0;
+    if (segments.size() > 1) {
+        size_t mid_size = 0;
+        size_t ix_last = segments.size() - 1;
+        size_t left_size = segments.front()->size_bytes() - left_scan_bytes;
+        size_t right_size = right_scan_bytes;
+        for (size_t i = 1; i < ix_last; i++) {
+            mid_size += segments[i]->size_bytes();
+        }
+        total_size = left_size + mid_size + right_size;
+        vlog(
+          stlog.debug,
+          "Computed size components: {}-{}-{}, total: {}, left scan bytes: {}, "
+          "right scan bytes: {}, num segments: {}, "
+          "offset range: {}-{}",
+          left_size,
+          mid_size,
+          right_size,
+          total_size,
+          left_scan_bytes,
+          right_scan_bytes,
+          segments.size(),
+          first,
+          last);
+    } else {
+        // Both left and right scans were performed on the same segment.
+        total_size = right_scan_bytes - left_scan_bytes;
+        vlog(
+          stlog.debug,
+          "Computed size total: {}, left scan bytes: {}, right scan bytes: {}, "
+          "num segments: {}, offset range: {}-{}",
+          total_size,
+          left_scan_bytes,
+          right_scan_bytes,
+          segments.size(),
+          first,
+          last);
+    }
+    co_return offset_range_size_result_t{
+      .on_disk_size = total_size,
+      .last_offset = last,
+    };
+}
+
+ss::future<std::optional<log::offset_range_size_result_t>>
+disk_log_impl::offset_range_size(
+  model::offset first,
+  offset_range_size_requirements_t target,
+  ss::io_priority_class io_priority) {
+    vlog(
+      stlog.debug,
+      "Offset range size, first: {}, target size: {}/{}, lstat: {}",
+      first,
+      target.target_size,
+      target.min_size,
+      offsets());
+    auto base_it = _segs.lower_bound(first);
+
+    // Invariant: 'first' offset should be present in the log. If the segment is
+    // compacted it's OK if it's missing if the segment that stores the offset
+    // range that includes the offset still exists.
+    if (base_it == _segs.end()) {
+        vlog(
+          stlog.debug,
+          "Offset {} is not present in the log {}",
+          first,
+          offsets());
+        co_return std::nullopt;
+    } else {
+        auto lstat = base_it->get()->offsets();
+        if (first < lstat.base_offset || first > lstat.committed_offset) {
+            vlog(
+              stlog.debug,
+              "Offset {} is not present in the segment {}",
+              first,
+              lstat);
+            co_return std::nullopt;
+        }
+    }
+
+    size_t first_segment_file_pos = 0;
+    auto first_segment = *base_it;
+    size_t first_segment_size = first_segment->file_size();
+    auto first_segment_offsets = first_segment->offsets();
+    auto r_lock = co_await first_segment->read_lock();
```

Comment:
In case it's unintentional, this lock isn't dropped until the end of this method